### PR TITLE
Assign4

### PR DESCRIPTION
**Pull request by aarushi121**

### CODE REVIEW

File: Assignment4/Assignment4.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
+var fs = require('fs');
+console.log("------------------------------------------");
+function ReadAppendWrite (read_file, append_file, write_file) {
```
Comment (rockon1985) [inline Assignment4/Assignment4.js]: - variable or parameter names must be _camelCased_ , not _underscore_separated_
- function name not correctly _camelCased_

File: Assignment4/Assignment4.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    fs.appendFile(append_file, data1, function(err2) {
+      if (err2)  throw err2;
+  
+      fs.readFile (append_file, function(err3, data2) {
+        if (err3)  throw err3;
+        
+        fs.writeFile(write_file, data2, function(err3) {
+          if (err2)  throw err2;
+        console.log(data2.toString());
+        console.log("-----------------------------------------------");
```
Comment (rockon1985) [inline Assignment4/Assignment4.js]: correct indentation here

File: Assignment4/Assignment4.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        });
+      });
+    });
+  });
+}
+ReadAppendWrite("file1.txt", "file2.txt", "file3.txt");
```
Comment (rockon1985) [inline Assignment4/Assignment4.js]: add new line at the end of each file

File: Assignment4/Assignment4.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
+var fs = require('fs');
```
Comment (rockon1985) [inline Assignment4/Assignment4.js]: use `const` instead of `var`

File: Assignment4/Assignment4.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+console.log("------------------------------------------");
+function ReadAppendWrite (read_file, append_file, write_file) {
+  fs.readFile (read_file, function(err1, data1) {
+    if (err1) throw err1;
+    console.log(data1.toString())
+    console.log("----------------------------------------------");
+
+    fs.appendFile(append_file, data1, function(err2) {
+      if (err2)  throw err2;
```
Comment (rockon1985) [inline Assignment4/Assignment4.js]: two spaces between if() and throw clause. One space required. Correct rest of the places as well

**Review by rockon1985 — CHANGES_REQUESTED** [review]
Make code styling changes.

**Review by rockon1985 — APPROVED** [review]
Good Job! Approved from my end!

File: Assignment4/Assignment4.js [hunk 6]
```diff
@@ -1,24 +1,24 @@
...
-        if (err3)  throw err3;
+      fs.readFile (appendFile, function (err3, data2) {
+        if (err3) throw err3;
         
-        fs.writeFile(write_file, data2, function(err3) {
-          if (err2)  throw err2;
-        console.log(data2.toString());
-        console.log("-----------------------------------------------");
+        fs.writeFile(writeFile, data2, function (err3) {
+          if (err2) throw err2;
```
Comment (sakodiya) [inline Assignment4/Assignment4.js]: @aarushi121 Please make separate function for throwing an error & writing console messages.

**Review by sakodiya — CHANGES_REQUESTED** [review]

File: Assignment4/Assignment4.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
+const fs = require('fs');
+
+function readAppendWrite (readFile, appendFile, writeFile) {
```
Comment (devanshutechracers) [inline Assignment4/Assignment4.js]: `readAppendWrite (readFile, appendFile, writeFile)` should be `readAppendWrite(readFile, appendFile, writeFile)`.
Problem Type: Spacing

File: Assignment4/Assignment4.js [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  fs.readFile (readFile, function (err1, data1) {
```
Comment (devanshutechracers) [inline Assignment4/Assignment4.js]: `fs.readFile (readFile, function (err1, data1)` should be `fs.readFile(readFile, function (err1, data1)`.
Problem Type: Spacing

Comment (devanshutechracers) [inline Assignment4/Assignment4.js]: `err1` and `data1`  can be a bit descritive, like `errorOnFile1Read`, `dataFromFile1` and similarly for other error and data parameter.
Problem Type: Naming Convention

File: Assignment4/Assignment4.js [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    console.log("----------------------------------------------");
+
+    fs.appendFile(appendFile, data1, function (err2) {
+      if (err2) throw err2;
+  
+      fs.readFile (appendFile, function (err3, data2) {
+        if (err3) throw err3;
+        
+        fs.writeFile(writeFile, data2, function (err3) {
+          if (err2) throw err2;
```
Comment (devanshutechracers) [inline Assignment4/Assignment4.js]: Throwing err2 on raising err3 error. Any logic behind this, which we should know?
Problem Type: Typo Error

**Review by devanshutechracers — CHANGES_REQUESTED** [review]
Check the Spacing/Naming Convention and a Typo Error. I have made the comments.

File: Assignment4/Assignment4.js [hunk 10]
```diff
@@ -0,0 +1,34 @@
+const fs = require('fs');
+
+function checkError(error){
```
Comment (devanshutechracers) [inline Assignment4/Assignment4.js]: space before starting block is missing.
Problem Type: Spacing

File: Assignment4/Assignment4.js [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+  if(error)
+    throw error;
+}
+
+function print(string)
+{
```
Comment (devanshutechracers) [inline Assignment4/Assignment4.js]: Block to be started in same line as function declaration.
Problem Type: Linting

File: Assignment4/Assignment4.js [hunk 12]
```diff
@@ -0,0 +1,34 @@
+const fs = require('fs');
+
+function checkError(error){
+  if(error)
```
Comment (devanshutechracers) [inline Assignment4/Assignment4.js]: space after `if` statement.
Problem Type: Spacing

File: Assignment4/Assignment4.js [hunk 13]
```diff
@@ -1,21 +1,31 @@
 const fs = require('fs');
 
-function readAppendWrite (readFile, appendFile, writeFile) {
-  fs.readFile (readFile, function (err1, data1) {
-    if (err1) throw err1;
-    console.log(data1.toString())
-    console.log("----------------------------------------------");
+function checkError(error){
```
Comment (sakodiya) [inline Assignment4/Assignment4.js]: @aarushi121 We can make only one function for error-throwing & printing console messages. Think about it.

File: Assignment4/Assignment4.js [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+    throw error;
+}
+
+function print(string)
+{
+  console.log(string);
+}
+
+function readAppendWrite(readFile, appendFile, writeFile) {
+  fs.readFile (readFile, function (errorOnFile1Read, dataFromFile1) {
```
Comment (devanshutechracers) [inline Assignment4/Assignment4.js]: No Space after function calling and before its argument. This should be like `fs.readFile(read......`
Problem Type: Spacing

File: Assignment4/Assignment4.js [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+    checkError (errorOnFile1Read);
```
Comment (devanshutechracers) [inline Assignment4/Assignment4.js]: Same here.
Problem Type: Spacing

File: Assignment4/Assignment4.js [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+
+function print(string)
+{
+  console.log(string);
+}
+
+function readAppendWrite(readFile, appendFile, writeFile) {
+  fs.readFile (readFile, function (errorOnFile1Read, dataFromFile1) {
+    checkError (errorOnFile1Read); 
+    print(dataFromFile1.toString())
```
Comment (devanshutechracers) [inline Assignment4/Assignment4.js]: Semicolon when line ends.
Problem Type: Missing Semicolon

File: Assignment4/Assignment4.js [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+    print("----------------------------------------------");
+
+    fs.appendFile(appendFile, dataFromFile1, function (errorOnFile2Append) {
+      checkError(errorOnFile2Append);
+  
+      fs.readFile(appendFile, function (errorOnFile2Read, dataFromFile2) {
+        checkError(errorOnFile2Read);
+        
+        fs.writeFile(writeFile, datafromFile2, function (errorOnFile3Write) {
+          checkError(errorOnFile3Write)
```
Comment (devanshutechracers) [inline Assignment4/Assignment4.js]: Semicolon when line ends.
Problem Type: Missing Semicolon

**Review by devanshutechracers — CHANGES_REQUESTED** [review]
Spaces and Semicolon Syntax to improve

**Review by sakodiya — APPROVED** [review]

File: Assignment4/Assignment4.js [hunk 18]
```diff
@@ -0,0 +1,35 @@
+const fs = require('fs');
+
+function checkError(error) {
+  if(error) {
```
Comment (devanshutechracers) [inline Assignment4/Assignment4.js]: space after `if` statement.
Problem Type: Spacing
Example:
`if (error) {` and not `if(error) {`

**Review by devanshutechracers — CHANGES_REQUESTED** [review]
Fix Space after if

File: Assignment4/Assignment4.js [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+    print(dataFromFile1.toString());
+    print("----------------------------------------------");
+
+    fs.appendFile(appendFile, dataFromFile1, function (errorOnFile2Append) {
+      checkError(errorOnFile2Append);
+  
+      fs.readFile(appendFile, function (errorOnFile2Read, dataFromFile2) {
+        checkError(errorOnFile2Read);
+        
+        fs.writeFile(writeFile, datafromFile2, function (errorOnFile3Write) {
```
Comment (devanshutechracers) [inline Assignment4/Assignment4.js]: `datafromFile2` is not defined.

Read the detail of the error below.

-----
fs.writeFile(writeFile, datafromFile2, function (errorOnFile3Write) {

ReferenceError: datafromFile2 is not defined
    at /home/rails/work/training/aarushi_goyal/node-training/Assignment4/Assignment4.js:26:33
    at FSReqWrap.readFileAfterClose [as oncomplete] (fs.js:445:3)

-----

File: Assignment4/Assignment4.js [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+          checkError(errorOnFile3Write);
+          print(data2.toString());
```
Comment (devanshutechracers) [inline Assignment4/Assignment4.js]: `data2` is not defined.

Read the detail of the error below.

------
print(data2.toString());

ReferenceError: data2 is not defined
    at /home/rails/work/training/aarushi_goyal/node-training/Assignment4/Assignment4.js:28:17
    at FSReqWrap.oncomplete (fs.js:123:15)

------

**Review by devanshutechracers — CHANGES_REQUESTED** [review]
Run Time Errors Observed while Test Run.

**Review by devanshutechracers — APPROVED** [review]
Approved.

**Pull request merged**